Add caching

marcelgarus opened the pull request:
Fixes #52 
Fixes #57 
Fixes #58 
Fixes #59 
Fixes #74 
Fixes #76 
Fixes #79 
Fixes #82 
Fixes #83 

* The `repository` package got replaced by the `streaming_shared_preferences` package for simple app-level value storage and the `cached_listview` package to cache module-level collections.
* Caching (and offline-access) and reloading via pull-to-refresh is now supported in the news, course and assignments module.
* We use the correct font.
* A new ColorSwatch is defined for the app.
* New widgets! 🎉 `PrimaryButton`, `SecondaryButton` (outlined), `EmptyStateScreen` and `ErrorScreen`

marcelgarus commented:
Travis build fails because of an issue in the build_runner framework: https://github.com/dart-lang/sdk/issues/35437

marcelgarus commented:
Better not merge it yet, I still want to update to a new version of `flutter_cached_listview` (which was renamed to `flutter_cached`).

marcelgarus commented:
Here's how the new UI stuff looks like:

No connection error:
![image](https://user-images.githubusercontent.com/8601189/66723614-93a36f00-ee1b-11e9-8204-47d683769baa.png)

Not authenticated error: 
![image](https://user-images.githubusercontent.com/8601189/66723618-9bfbaa00-ee1b-11e9-9a54-5389891a82a6.png)

Button being tapped:
![image](https://user-images.githubusercontent.com/8601189/66723622-a8800280-ee1b-11e9-8c50-b2468d5584af.png)

Color palette with shades:
![image](https://user-images.githubusercontent.com/8601189/66723631-c6e5fe00-ee1b-11e9-9164-63e7f4215144.png)

Pulling to refresh:
![image](https://user-images.githubusercontent.com/8601189/66723661-03195e80-ee1c-11e9-839a-6775aea8bce2.png)

It's refreshing:
![image](https://user-images.githubusercontent.com/8601189/66723665-1298a780-ee1c-11e9-9676-30e553ce7988.png)

If an error occurred while fetching data (like, when there's no network), for now a red container is displayed at the top. This will be replaced with a fancy error banner later on (see #59):
![image](https://user-images.githubusercontent.com/8601189/66723678-3360fd00-ee1c-11e9-8288-5b1ce822a10f.png)

marcelgarus commented:
[Someone suggested](https://github.com/dart-lang/sdk/issues/35437#issuecomment-541458554), we could try upgrading the `analyzer` package to fix the Travis build. Will try that tomorrow.

marcelgarus commented:
Upgrading the `analyzer` actually fixed our problem, so this PR is ready to be reviewed. 🎉

marcelgarus commented:
Also fixed the login and while on it, revamped the login screen a little bit:
![image](https://user-images.githubusercontent.com/8601189/67627398-f1907780-f85b-11e9-89dc-f41d26ed13a4.png)

janrenz commented:
@JonasWanke Can you please review?

janrenz commented:
@ctiedt Of course you too

JonasWanke commented on lib/hive.dart:
```diff
@@ -1,12 +1,30 @@
 bool _isHiveInitialized = false;
 
+const cacheArticlesKey = 'articles';
+const cacheAssignmentsKey = 'sssignments';
```
Typo

JonasWanke replied:
Typo

JonasWanke commented on lib/hive.dart:
```diff
@@ -15,41 +33,46 @@ Future<void> initializeHive() async {
     ..registerAdapter(IdAdapter<Author>(), 46)
-    // App module
+    ..registerAdapter(IdAdapter<Assignment>(), 47)
+    ..registerAdapter(IdAdapter<File>(), 49)
```
Looks like these belong to the individual modules and not the general section

JonasWanke replied:
As these belong to individual modules they should be moved into their respective sections below.

JonasWanke commented on lib/app/utils.dart:
```diff
@@ -61,44 +68,64 @@ Future<bool> tryLaunchingUrl(String url) async {
 /// A special kind of item that also carries its id.
 abstract class Entity {
-  Id<Entity> get id;
+  Id get id;
```
Why isn't the `Id` typed?

JonasWanke replied:
Why isn't this typed?

JonasWanke commented on lib/app/services/storage.dart:
```diff
@@ -1,37 +1,26 @@
-  set email(String email) => data = data.copy((data) => data..email = email);
-  set token(String token) => data = data.copy((data) => data..token = token);
+    await cache.initialize();
+  }
```
An initialize method without checking if this service is already initialized seems unsafe. What about a private constructor and a public static async method creating an initialized instance?

JonasWanke replied:
This method seems unsafe (not checking if it's already been initialized). Can we instead use a private constructor + static async "factory"?

JonasWanke commented on lib/main.dart:
```diff
@@ -1,45 +1,48 @@
+  await initializeHive();
+  final storage = StorageService();
+  await storage.initialize();
+  await initializeCacheRootKeys(storage);
```
Can we move this call to the `StorageService`'s initializer?

JonasWanke commented on lib/app/services/network.dart:
```diff
@@ -1,14 +1,23 @@
+  Duration timeToWait;
+}
+
 /// A service that offers networking post and get requests to the backend
```
"POST" and "GET" (uppercase) would be more clear

JonasWanke commented on lib/app/services/network.dart:
```diff
@@ -31,13 +40,23 @@ class NetworkService {
       await _ensureConnectionExists();
       var response = await call('$apiUrl/$path');
 
+      // Succeed, if its a 2xx status code.
```
Are you sure this comma is required/allowed?

JonasWanke commented on lib/app/utils.dart:
```diff
@@ -61,44 +68,64 @@ Future<bool> tryLaunchingUrl(String url) async {
+  } catch (_) {
+    return defaultValue;
+  }
+}
```
This method is only used once and should be inlined into `NetworkService` because of its simplicity.

JonasWanke commented on lib/app/services/network.dart:
```diff
@@ -61,7 +81,8 @@ class NetworkService {
     if (parameters.isNotEmpty) {
-      path += '?' + parameters.keys.map((p) => '$p=${parameters[p]}').join('&');
+      path +=
+          '?' + parameters.entries.map((p) => '${p.key}=${p.value}').join('&');
```
The parameters should be URI-encoded

JonasWanke commented on lib/app/services/user_fetcher.dart:
```diff
@@ -0,0 +1,30 @@
+
+  UserFetcherService({@required this.network}) : assert(network != null);
+
+  Future<User> fetchUser(Id<User> id) async {
```
Missing `null`-check for `id`

JonasWanke commented on lib/app/services/user_fetcher.dart:
```diff
@@ -0,0 +1,30 @@
+
+  Future<User> fetchUser(Id<User> id) async {
+    var response = await network.get('users/$id');
+    var data = json.decode(response.body);
```
Is there a reason to use `var` over `final`?

JonasWanke commented on lib/app/services/user_fetcher.dart:
```diff
@@ -0,0 +1,30 @@
+      firstName: data['firstName'],
+      lastName: data['lastName'],
+      email: data['email'],
+      schoolToken: data['schoolId'],
```
Why is this called token locally?

JonasWanke commented on lib/app/services/me.dart:
```diff
@@ -1,57 +1,79 @@
+      await _initializer;
     }
+    assert(_me != null);
   }
```
Similar to `StorageService` an async "factory" method would be cleaner.

JonasWanke commented on lib/app/services/me.dart:
```diff
@@ -1,57 +1,79 @@
+  StorageService storage;
+
+  Box _me;
+  int _meKey = 0;
```
Looks like this could be a `static const`

JonasWanke commented on lib/app/services/me.dart:
```diff
@@ -1,57 +1,79 @@
+  UserFetcherService userFetcher;
+  StorageService storage;
+
+  Box _me;
```
Can we type this to `Box<User>`? Also, `_box` would be a more descriptive name.

JonasWanke commented on lib/app/services/me.dart:
```diff
@@ -1,57 +1,79 @@
+        if (me == null) {
+          throw Exception('Item not in cache.');
+        }
+        return me;
```
We can reduce this to `return _me.get(_meKey) ?? (throw Exception('...'));`

JonasWanke commented on lib/app/services/me.dart:
```diff
@@ -1,57 +1,79 @@
+        await _ensureInitialized();
+        final me = _me.get(_meKey) as User;
+        if (me == null) {
+          throw Exception('Item not in cache.');
```
Can we make the exception message more specific than "item"?

JonasWanke commented on lib/main.dart:
```diff
@@ -1,45 +1,48 @@
+          dispose: (_, me) => me.dispose(),
         ),
       ],
       child: SchulCloudApp(),
```
With regard to the single responsibility principle, I'd prefer passing this via the constructor instead of hardcoding it here.

JonasWanke commented on lib/app/widgets/menu.dart:
```diff
@@ -59,39 +53,50 @@ class Menu extends StatelessWidget {
+                builder: (context, snapshot) {
+                  // We don't need to do extensive error handling here,
+                  // because it shouldn't be possible to open the menu
+                  // without being logged in.
```
Then maybe log an error in this case?

JonasWanke commented on lib/app/widgets/buttons.dart:
```diff
@@ -0,0 +1,47 @@
+  final VoidCallback onPressed;
+  final Widget child;
+
+  const PrimaryButton({Key key, this.child, this.onPressed}) : super(key: key);
```
Different order of `child` and `onPressed`

JonasWanke replied:
The constructor should be defined before fields

JonasWanke commented on lib/app/utils.dart:
```diff
@@ -61,44 +68,64 @@ Future<bool> tryLaunchingUrl(String url) async {
+  await Provider.of<StorageService>(context).clear();
+  Navigator.of(context, rootNavigator: true).pushReplacement(TopLevelPageRoute(
+    builder: (_) => LoginScreen(),
+  ));
```
Maybe we should move this to the login-module, MeService or some other related place other than general utils?

MarcelGarus replied:
Hmm, I agree that the logic is login-related, so the `await Provider.of<StorageService>(context).clear();` could indeed be put into the `MeService`. 😊
But the `Navigator....` part is UI-logic and I'm not sure we should mix that into the domain-specific logic.
And because it's used in various places (like the `ErrorScreen` for a `NotAuthenticatedError` or the `Menu` for the logout button.

JonasWanke commented on lib/app/widgets/empty_state.dart:
```diff
@@ -0,0 +1,58 @@
+/// A screen with no content on it. Instead a placeholder is displayed,
+/// consisting of [child] (commonly an image) and a text underneath as well as
+/// some actions that the user can take.
+/// Trying to executed the last action again (like, fetching stuff from the
```
"Trying to _execute_..."

JonasWanke commented on lib/app/widgets/empty_state.dart:
```diff
@@ -0,0 +1,58 @@
+              LimitedBox(
+                maxHeight: 300,
+                child: SvgPicture.asset('assets/empty_states/default.svg'),
+              ),
```
Can we use `SvgPicture`'s `height` parameter instead?

JonasWanke commented on lib/app/widgets/error_banner.dart:
```diff
@@ -0,0 +1,43 @@
+  final dynamic error;
+  final VoidCallback onRetry;
+
+  ErrorBanner(this.error, {this.onRetry}) : assert(error != null);
```
Constructors should be defined before fields

JonasWanke commented on lib/app/widgets/empty_state.dart:
```diff
@@ -0,0 +1,58 @@
+    this.actions = const [],
+    this.onRetry,
+  })  : assert(text != null),
+        assert(actions != null);
```
Constructors should be defined before fields

JonasWanke commented on lib/app/widgets/error_banner.dart:
```diff
@@ -0,0 +1,43 @@
+        onPressed: () => logOut(context),
+      ));
+    } else {
+      message = "Oh no! An internal error occurred:\n$error";
```
Double quotes instead of single quotes

MarcelGarus replied:
According to the discussion at https://github.com/dart-lang/site-www/issues/859, you should use double quotes if there's a single quote inside the paragraph.
But of course, I'll change the quotes on line 30.

JonasWanke commented on lib/app/widgets/error_banner.dart:
```diff
@@ -0,0 +1,43 @@
+  @override
+  Widget build(BuildContext context) {
+    String message;
+    List<Widget> actions = [];
```
These actions aren't used anywhere

JonasWanke commented on lib/app/widgets/error_screen.dart:
```diff
@@ -0,0 +1,48 @@
+  final dynamic error;
+  final VoidCallback onRetry;
+
+  ErrorScreen(this.error, {this.onRetry}) : assert(error != null);
```
These actions aren't used anywhere

JonasWanke replied:
The constructor should be defined before fields

JonasWanke commented on lib/app/widgets/error_screen.dart:
```diff
@@ -0,0 +1,48 @@
+        onPressed: () => logOut(context),
+      ));
+    } else {
+      message = "Oh no! An internal error occurred:\n$error";
```
Double quotes instead of single quotes

JonasWanke commented on lib/app/widgets/error_screen.dart:
```diff
@@ -0,0 +1,48 @@
+        child: LimitedBox(
+          maxHeight: 300,
+          child: SvgPicture.asset('assets/empty_states/broken_pen.svg'),
+        ),
```
Can we use `SvgPicture`'s `height` parameter instead?

JonasWanke commented on lib/login/widgets/form.dart:
```diff
@@ -0,0 +1,138 @@
+      ));
+    } on NoConnectionToServerError catch (_) {
+      _ambientError = 'No connection to the server.';
+    } on AuthenticationError catch (_) {
```
You can omit `catch (_)`

JonasWanke commented on lib/login/widgets/login_screen.dart:
```diff
@@ -3,169 +3,90 @@ import 'package:provider/provider.dart';
+        child: CustomScrollView(
+          slivers: <Widget>[
+            SliverList(
+              delegate: SliverChildBuilderDelegate(_buildSliver),
```
A `SliverChildListDelegate` should be more readable and doesn't lag when scrolling. We can think of a solution similar to this one when we have content that takes more time to load

JonasWanke commented on lib/login/widgets/form.dart:
```diff
@@ -0,0 +1,138 @@
+    setState(() => _isLoading = true);
+
+    try {
+      await login();
+      setState(() => _ambientError = null);
+
+      // Logged in.
+      Navigator.of(context).pushReplacement(TopLevelPageRoute(
+        builder: (_) => LoggedInScreen(),
+      ));
```
This is being called even when validation fails

MarcelGarus replied:
Doesn't `login()` throw an `AuthenticationError` when the login failed?

JonasWanke replied:
Not necessarily. If validation falls, the BLoC's login method isn't called (see a few lines further down) so `login` returns normally

JonasWanke commented on lib/login/widgets/slanted_section.dart:
```diff
@@ -0,0 +1,76 @@
+import 'package:flutter/material.dart';
+
+enum SlantedSectionType { onlyTopSlanted, onlyBottomSlanted, bothSlanted }
```
Doesn't seem to be used anywhere

JonasWanke commented on lib/login/widgets/slanted_section.dart:
```diff
@@ -0,0 +1,76 @@
+    return oldDelegate is SlantedSectionPainter &&
+        oldDelegate.color == color &&
+        oldDelegate.slantTop == slantTop &&
+        oldDelegate.slantBottom == slantBottom;
```
I think these conditions should be inverted — we should repaint if some property _changed_ and not if everything stays the same

JonasWanke commented on lib/assignments/bloc.dart:
```diff
@@ -0,0 +1,95 @@
+          parentKey: cacheAssignmentsKey,
+          fetcher: () async {
+            var response = await network.get('homework');
+            var body = json.decode(response.body);
```
Can both be final

JonasWanke commented on lib/assignments/bloc.dart:
```diff
@@ -0,0 +1,95 @@
+
+            return [
+              for (var data in body['data'] as List<dynamic>)
+                Assignment(
```
Can we use `List.map(...)` for better readability?

MarcelGarus replied:
It's still readable as-is and we asynchronously fetch the user inside the `Assignment` constructor, so we would have to do stuff with `Future.wait`, because a plain `List` doesn't support `asyncMap` for now... 😕

JonasWanke commented on lib/assignments/data.dart:
```diff
@@ -1,16 +1,15 @@
 
 @HiveType()
-class Homework implements Entity {
+class Assignment implements Entity, Comparable {
```
`@immutable` missing

JonasWanke commented on lib/assignments/data.dart:
```diff
@@ -87,7 +91,7 @@ class Submission implements Entity {
 
   @HiveField(2)
-  final Id<Homework> homeworkId;
+  final Id<Assignment> homeworkId;
```
Should be renamed to `assignmentId`

JonasWanke commented on lib/assignments/data.dart:
```diff
@@ -76,6 +75,11 @@ class Homework implements Entity {
+  }
 }
 
 @HiveType()
```
`@immutable` missing

JonasWanke commented on lib/assignments/bloc.dart:
```diff
@@ -0,0 +1,95 @@
+          parentKey: cacheSubmissionsKey,
+          fetcher: () async {
+            var response = await network.get('submissions');
+            var body = json.decode(response.body);
```
Can be final

JonasWanke commented on lib/assignments/bloc.dart:
```diff
@@ -0,0 +1,95 @@
+
+            return [
+              for (var data in body['data'] as List<dynamic>)
+                Submission(
```
Can we use `List.map(...)` for better readability?

JonasWanke commented on lib/assignments/bloc.dart:
```diff
@@ -0,0 +1,95 @@
+                        await userFetcher.fetchUser(Id<User>(id)),
+                    ],
+                    color: hexStringToColor(data['courseId']['color']),
+                  ),
```
Can we just use the `courseId` instead of parsing the course? We can then normalize the data by looking up the course individually.

JonasWanke commented on lib/assignments/bloc.dart:
```diff
@@ -0,0 +1,95 @@
+                        'No description provided',
+                    teachers: [
+                      for (String id in data['courseId']['teacherIds'])
+                        await userFetcher.fetchUser(Id<User>(id)),
```
Seems very slow to explicitly load teachers from the API without caching them locally

JonasWanke commented on lib/assignments/bloc.dart:
```diff
@@ -0,0 +1,95 @@
+  }
+
+  Stream<Submission> getSubmissionForHomework(
+      Id<Assignment> homeworkId) async* {
```
Assertion for `null` `homeworkId` missing (and it should be called `assignmentId`)

JonasWanke commented on lib/assignments/bloc.dart:
```diff
@@ -0,0 +1,95 @@
+              (submission) => submission.homeworkId == homeworkId,
+              orElse: () => null,
+            ))
+        .firstWhere((submission) => submission != null);
```
It seems unnecessary to fetch all submissions first and then filter. Can't we directly fetch by ID?

MarcelGarus replied:
The API doesn't support that. Assignments (which are called "homework" in the backend (but not in the web client UI)) don't reference a "submissionId" and according to [the api docs](https://api.schul-cloud.org/docs/#/submissions), submissions don't seem to support fetching based on an assignment ID

JonasWanke replied:
What about getting a single one from disk directly and only fetching all when accessing the API?

JonasWanke commented on lib/assignments/widgets/assignment_details_screen.dart:
```diff
@@ -7,15 +7,15 @@ import '../bloc.dart';
+  final Assignment homework;
 
-  const HomeworkDetailScreen({Key key, @required this.homework})
+  const AssignmentDetailsScreen({Key key, @required this.homework})
```
- `null` assertion for `homework` missing
- called `homework` instead of `assignment`

JonasWanke replied:
The constructor should be defined before fields

JonasWanke commented on lib/assignments/widgets/assignment_details_screen.dart:
```diff
@@ -7,15 +7,15 @@ import '../bloc.dart';
   void _showSubmissionScreen(
     BuildContext context,
-    Homework homework,
+    Assignment homework,
```
`homework` instead of `assignment`

JonasWanke commented on lib/assignments/widgets/submission_screen.dart:
```diff
@@ -4,7 +4,7 @@ import 'package:flutter_html/flutter_html.dart';
 
 class SubmissionScreen extends StatelessWidget {
-  final Homework homework;
+  final Assignment homework;
```
`homework` instead of `assignment`

JonasWanke commented on lib/assignments/widgets/assignments_screen.dart:
```diff
@@ -0,0 +1,109 @@
+import '../data.dart';
+import 'assignment_details_screen.dart';
+
+class HomeworkScreen extends StatelessWidget {
```
`homework` instead of `assignment`

JonasWanke commented on lib/assignments/widgets/assignments_screen.dart:
```diff
@@ -0,0 +1,109 @@
+                      DateTime(h.dueDate.year, h.dueDate.month, h.dueDate.day),
+                );
+
+                var dates = assignments.keys.toList()
```
Can be final

JonasWanke commented on lib/assignments/widgets/assignments_screen.dart:
```diff
@@ -0,0 +1,109 @@
+            return CachedBuilder(
+              controller: bloc.assignments,
+              errorBannerBuilder: (_, error) => ErrorBanner(error),
+              errorScreenBuilder: (_, error) => ErrorScreen(error),
```
Maybe we can extract this into a new widget

JonasWanke commented on lib/assignments/widgets/assignments_screen.dart:
```diff
@@ -0,0 +1,109 @@
+}
+
+class AssignmentCard extends StatelessWidget {
+  final Assignment homework;
```
`homework` instead of `assignment`

JonasWanke commented on lib/courses/bloc.dart:
```diff
@@ -1,80 +1,80 @@
+  final _lessons = <Id<Course>, CacheController<List<Lesson>>>{};
+
+  Bloc({
+    @required this.storage,
```
assertion missing

JonasWanke commented on lib/courses/bloc.dart:
```diff
@@ -1,80 +1,80 @@
+          parentKey: cacheCoursesKey,
+          fetcher: () async {
+            var response = await network.get('courses');
+            var body = json.decode(response.body);
```
Can be final

JonasWanke commented on lib/courses/bloc.dart:
```diff
@@ -1,80 +1,80 @@
+
+            return [
+              for (var data in body['data'] as List<dynamic>)
+                Course(
```
Can we use `List.map(...)` instead?

JonasWanke commented on lib/courses/bloc.dart:
```diff
@@ -1,80 +1,80 @@
+                  description: data['description'],
+                  teachers: [
+                    for (String id in data['teacherIds'])
+                      await userFetcher.fetchUser(Id<User>(id)),
```
Explicit fetching without cache

JonasWanke commented on lib/courses/data.dart:
```diff
@@ -50,7 +49,7 @@ class Content implements Entity {
 @immutable
 @HiveType()
-class Course implements Entity {
+class Course implements Entity, Comparable {
```
`teachers` could be normalized to IDs (if hive doesn't do this already)

JonasWanke replied:
The constructor should be defined before fields

JonasWanke commented on lib/courses/bloc.dart:
```diff
@@ -1,80 +1,80 @@
+    courses.dispose();
+    for (var lesson in _lessons.values) {
+      lesson.dispose();
+    }
```
I think `foreach` is more readable

JonasWanke commented on lib/courses/bloc.dart:
```diff
@@ -1,80 +1,80 @@
+              parentKey: courseId.id,
+              fetcher: () async {
+                var response = await network.get('lessons?courseId=$courseId');
+                var body = json.decode(response.body);
```
Can be final

JonasWanke commented on lib/courses/bloc.dart:
```diff
@@ -1,80 +1,80 @@
+
+                return [
+                  for (var data in body['data'] as List<dynamic>)
+                    Lesson(
```
Can we use `List.map(...)` instead?

JonasWanke commented on lib/courses/widgets/course_detail_screen.dart:
```diff
@@ -35,55 +44,51 @@ class CourseDetailScreen extends StatelessWidget {
+                builder: (BuildContext context, List<Lesson> lessons) {
+                  if (lessons.isEmpty) {
+                    return EmptyStateScreen(
+                      text: "Seems like you're not enrolled in any courses.",
```
Double quotes instead of single quotes

MarcelGarus replied:
"you're" would otherwise have to be escaped

JonasWanke commented on lib/courses/widgets/course_detail_screen.dart:
```diff
@@ -35,55 +44,51 @@ class CourseDetailScreen extends StatelessWidget {
-          child: _LessonList(course: course),
+          child: Consumer<Bloc>(
+            builder: (context, bloc, _) {
+              return CachedBuilder(
```
Can be an expression body

MarcelGarus replied:
According to the [Effective Dart guide, we shouldn't do that](https://dart.dev/guides/language/effective-dart/usage#consider-using--for-simple-members), because our declaration is more than a couple of lines and contains deeply nested expression.

JonasWanke commented on lib/courses/widgets/course_detail_screen.dart:
```diff
@@ -7,21 +8,29 @@ import '../bloc.dart';
+  }
+
   @override
   Widget build(BuildContext context) {
```
The `build`-method is very long. Maybe we can extract e.g. the `CachedBuilder`?

JonasWanke commented on lib/courses/widgets/courses_screen.dart:
```diff
@@ -9,33 +10,36 @@ import 'course_card.dart';
+                if (courses.isEmpty) {
+                  return EmptyStateScreen(
+                    text: "Seems like you're currently not enrolled in any "
+                        "courses.",
```
Double quotes i/o single quotes

JonasWanke commented on lib/news/bloc.dart:
```diff
@@ -1,52 +1,41 @@
+
+            return [
+              for (var data in body['data'] as List<dynamic>)
+                Article(
```
Can we use `List.map(...)` instead?

JonasWanke commented on lib/news/bloc.dart:
```diff
@@ -1,52 +1,41 @@
+          parentKey: cacheArticlesKey,
+          fetcher: () async {
+            var response = await network.get('news?');
+            var body = json.decode(response.body);
```
Can be final

JonasWanke commented on lib/news/bloc.dart:
```diff
@@ -1,52 +1,41 @@
+                    id: Id<Author>(data['creator']['_id']),
+                    name:
+                        '${data['creator']['firstName']} ${data['creator']['lastName']}',
+                  ),
```
Looks like this could be stored as a user ID. Just as it's implemented in the line above. So simply remove the `author`-field.

JonasWanke commented on lib/news/bloc.dart:
```diff
@@ -1,52 +1,41 @@
+                    name:
+                        '${data['creator']['firstName']} ${data['creator']['lastName']}',
+                  ),
+                  section: 'Section',
```
Isn't used yet so at least hide it

JonasWanke commented on lib/news/widgets/news_screen.dart:
```diff
@@ -10,36 +11,32 @@ import 'article_preview.dart';
+                return ListView.builder(
+                  itemCount: articles.length,
+                  itemBuilder: (context, index) {
+                    var article = articles[index];
```
Can be final — but do we even need this as a separate variable?

JonasWanke commented on lib/news/widgets/news_screen.dart:
```diff
@@ -10,36 +11,32 @@ import 'article_preview.dart';
+      child: Scaffold(
+        body: Consumer<Bloc>(
+          builder: (context, bloc, _) {
+            return CachedBuilder<List<Article>>(
```
Can be an expression body

MarcelGarus replied:
https://dart.dev/guides/language/effective-dart/usage#consider-using--for-simple-members

JonasWanke commented on lib/file_browser/bloc.dart:
```diff
@@ -1,29 +1,53 @@
+            (body as List<dynamic>).where((file) => file['name'] != null);
+        return [
+          for (final data in validFiles)
+            File(
```
Can we use `List.map(...)` instead?

JonasWanke commented on lib/file_browser/data.dart:
```diff
@@ -1,26 +1,35 @@
+part 'data.g.dart';
+
+@HiveType()
+class File implements Entity, Comparable {
```
`@immutable` missing

JonasWanke replied:
The constructor should be defined before fields

JonasWanke commented on lib/courses/widgets/files_screen.dart:
```diff
@@ -74,28 +77,34 @@ class _CourseFilesList extends StatelessWidget {
+            errorBannerBuilder: (_, error) =>
+                Container(color: Colors.red, height: 48),
+            errorScreenBuilder: (_, error) => ErrorScreen(error),
+            builder: (BuildContext context, List<Course> courses) {
```
Can be an expression body

JonasWanke commented on lib/courses/widgets/files_screen.dart:
```diff
@@ -74,28 +77,34 @@ class _CourseFilesList extends StatelessWidget {
+              return ListView(
+                children: <Widget>[
+                  for (var course in courses)
+                    ListTile(
```
Can we use `List.map(...)` instead?

JonasWanke commented on lib/app/utils.dart:
```diff
@@ -61,44 +68,64 @@ Future<bool> tryLaunchingUrl(String url) async {
+
 /// An error indicating that a permission wasn't granted by the user.
 class PermissionNotGranted<T> implements Exception {
   String toString() => "A permission wasn't granted by the user.";
```
Double quotes i/o single quotes

JonasWanke commented on lib/app/utils.dart:
```diff
@@ -61,44 +68,64 @@ Future<bool> tryLaunchingUrl(String url) async {
+          fetcher: () async {
+            return await fetcher()
+              ..sort();
+          },
```
Can be expression bodies

JonasWanke commented on lib/app/utils.dart:
```diff
@@ -61,44 +68,64 @@ Future<bool> tryLaunchingUrl(String url) async {
+              ..sort();
+          },
+          fetcher: () async {
+            return await fetcher()
```
`fetcher` can be null

JonasWanke commented on lib/app/services/me.dart:
```diff
@@ -1,57 +1,79 @@
+
+  User get me => meController.lastData;
+
+  MeService({
```
The constructor should be defined before fields

JonasWanke commented on lib/app/services/storage.dart:
```diff
@@ -1,37 +1,26 @@
-    await _inMemory.update(_dataId, StorageData());
-    await _storage.loadItemsIntoCache();
-  }
+  StorageService() : cache = HiveCache();
```
The constructor should be defined before fields

JonasWanke commented on lib/app/services/user_fetcher.dart:
```diff
@@ -0,0 +1,30 @@
+class UserFetcherService {
+  final NetworkService network;
+
+  UserFetcherService({@required this.network}) : assert(network != null);
```
The constructor should be defined before fields

JonasWanke commented on lib/app/utils.dart:
```diff
@@ -61,44 +68,64 @@ Future<bool> tryLaunchingUrl(String url) async {
+class Id<T> {
+  final String id;
+
+  Id(this.id);
```
The constructor should be defined before fields

JonasWanke commented on lib/app/utils.dart:
```diff
@@ -61,44 +68,64 @@ Future<bool> tryLaunchingUrl(String url) async {
 abstract class Entity {
-  Id<Entity> get id;
+  Id get id;
   const Entity();
```
The constructor should be defined before fields

JonasWanke commented on lib/app/utils.dart:
```diff
@@ -61,44 +68,64 @@ Future<bool> tryLaunchingUrl(String url) async {
+  final StorageService storage;
+  final String parentKey;
+
+  HiveCacheController({
```
The constructor should be defined before fields

JonasWanke commented on lib/assignments/bloc.dart:
```diff
@@ -0,0 +1,95 @@
+  CacheController<List<Assignment>> assignments;
+  CacheController<List<Submission>> submissions;
+
+  Bloc({
```
The constructor should be defined before fields

JonasWanke commented on lib/assignments/data.dart:
```diff
@@ -54,7 +53,7 @@ class Homework implements Entity {
   final int maxTeamMembers;
 
-  Homework({
+  Assignment({
```
The constructor should be defined before fields

JonasWanke commented on lib/assignments/data.dart:
```diff
@@ -87,7 +91,7 @@ class Submission implements Entity {
   final String schoolId;
```
The constructor should be defined before fields

JonasWanke commented on lib/assignments/widgets/submission_screen.dart:
```diff
@@ -4,7 +4,7 @@ import 'package:flutter_html/flutter_html.dart';
+  final Assignment homework;
   final Submission submission;
 
   const SubmissionScreen({
```
The constructor should be defined before fields

JonasWanke commented on lib/courses/bloc.dart:
```diff
@@ -1,80 +1,80 @@
+  CacheController<List<Course>> courses;
+  final _lessons = <Id<Course>, CacheController<List<Lesson>>>{};
+
+  Bloc({
```
The constructor should be defined before fields

JonasWanke commented on lib/courses/widgets/course_detail_screen.dart:
```diff
@@ -7,21 +8,29 @@ import '../bloc.dart';
   final Course course;
 
-  CourseDetailScreen({@required this.course}) : assert(course != null);
+  CourseDetailsScreen({@required this.course}) : assert(course != null);
```
The constructor should be defined before fields

JonasWanke commented on lib/file_browser/bloc.dart:
```diff
@@ -1,29 +1,53 @@
-  Repository<File> _files;
+  CacheController<List<File>> files;
 
   Bloc({
```
The constructor should be defined before fields

JonasWanke commented on lib/file_browser/widgets/file_browser.dart:
```diff
@@ -11,14 +11,6 @@ import '../data.dart';
-  ),
-);
-
 class FileBrowser extends StatelessWidget {
```
The constructor should be defined before fields

JonasWanke commented on lib/login/widgets/input.dart:
```diff
@@ -0,0 +1,32 @@
+  final bool obscureText;
+  final VoidCallback onChanged;
+
+  LoginInput({
```
The constructor should be defined before fields

JonasWanke commented on lib/main.dart:
```diff
@@ -1,45 +1,48 @@
 class ServicesProvider extends StatelessWidget {
+  final StorageService storage;
+
+  ServicesProvider({@required this.storage}) : assert(storage != null);
```
The constructor should be defined before fields

JonasWanke commented on lib/news/bloc.dart:
```diff
@@ -1,52 +1,41 @@
-          cache: HiveRepository<Article>('articles'),
+  CacheController<List<Article>> articles;
+
+  Bloc({@required StorageService storage, @required NetworkService network})
```
The constructor should be defined before fields

JonasWanke commented on lib/news/data.dart:
```diff
@@ -1,13 +1,12 @@
 @immutable
 @HiveType()
-class Article implements Entity {
+class Article implements Entity, Comparable {
```
The constructor should be defined before fields

JonasWanke commented on lib/app/utils.dart:
```diff
@@ -61,44 +68,64 @@ Future<bool> tryLaunchingUrl(String url) async {
+}
 
-  Future<List<Item>> downloadAll();
+Future<void> logOut(BuildContext context) async {
```
logout doesn't clear history, i.e. pressing the back button after logout navigates back to the dashboard/etc.

marcelgarus commented:
I implemented the requested changes.

About the general remarks:

* The button on the login screen is now named `MorphingLoadingButton`.
* The empty state is used if there are no courses.
* Double-checked the logo and looks just like [the original](https://github.com/schul-cloud/schulcloud-client/blob/develop/static/images/logo/HPI-Schul-Cloud_Wort-Bildmarke-Banner_transparent.svg)
* Cache layer now persists the item order.
* Local data was actually cleared before.
* Moved the HTML tag removal issue into a [separate issue](https://github.com/schul-cloud/schulcloud-flutter/issues/87) in order to prevent feature creep at this issue (or, at least, limit it🙄).
* File-related stuff is now all inside the `file` module.
* Speaking of modules, they now all have shorter singular names. (Except `news`, because the singular means a different thing.)

Regarding the observed issues:

* I can't reproduce the error when changing device orientation on the dashboard.
* Trying to log in with "a"/"" combination doesn't work for me and displays errors beneath the input fields, just as it should.

Also looking forward to round 2! 😄

janrenz commented:
Proposal: get this merge and make follow up tickets for remaining things?
